### PR TITLE
Validate Astro frontmatter JS/TS on compiler error

### PR DESCRIPTION
**Pull request by FredKSchott**
## Changes

- Before passing a component to the compiler, use esbuild to validate the frontmatter against JS/TS syntax.
- This will prevent a class of bug where bad front-matter syntax can break the compiled output in very obscure ways (for example, an unclosed parenthesis).
- Once the compiler can handle this better, we can remove this.

### Before
```
---
import Layout from '../layouts/MainLayout.astro';
{
--- 
```

Unclear error output.

Screen Shot 2021-12-03 at 11 58 30 AM

### After

Clear error output, with the correct bug location. "end of file" is still a little odd wording (it should say "end of frontmatter", which we can fix by re-writing this error message if that makes sense) but this is already a nice improvement over broken output, compiler panics, etc.

Screen Shot 2021-12-03 at 11 58 49 AM

## Testing

- Still TODO

## Docs

- Not needed

### CODE REVIEW

File: packages/astro/src/vite-plugin-astro/index.ts [hunk 1]
```diff
@@ -88,6 +97,7 @@ export default function astro({ config, devServer }: AstroPluginOptions): vite.P
         if (cssTransformError) throw cssTransformError;
 
         // Compile `.ts` to `.js`
+        console.log(tsResult.code);
```
Comment (FredKSchott) [inline packages/astro/src/vite-plugin-astro/index.ts:100]: doh, this should be removed. Thanks lint-bot!

**Review by drwpow — APPROVED** [review]
Pretty clever! This makes a lot of sense to me. And good error output

**Comment by matthewp** [comment]
Can we do this in the `catch` instead? That way the cost is only occurred on code that has an error anyways and not for every compile that we do.

**Comment by FredKSchott** [comment]
@matthewp clever! yes, that makes a ton of sense. 

I'll make that change, clean the PR up, and some tests, and submit for re-review.

**Review by natemoo-re — COMMENTED** [review]
On board with the concept, have reservations about this apporach. 

Since the Astro compiler already does a lexical scan, what if it exposed a utility to output the raw frontmatter? We could then validate it separately from the rest of the generated file.

**Comment by FredKSchott** [comment]
I'll also rewrite "unexpected end of file" to "unexpected end of frontmatter" if there are no objections.

File: packages/astro/src/vite-plugin-astro/index.ts [hunk 2]
```diff
@@ -51,6 +51,15 @@ export default function astro({ config, devServer }: AstroPluginOptions): vite.P
       let tsResult: TransformResult | undefined;
       let cssTransformError: Error | undefined;
 
+      // Precheck: verify that the frontmatter is valid TS before sending to our compiler.
+      // The compiler doesn't currently validate the frontmatter, so invalid JS/TS can cause issues
+      // in the final output. This prevents bad JS from becoming a compiler panic or otherwise
+      // breaking the final output.
+      const scannedFrontmatter = /^\-\-\-(.*)^\-\-\-/ms.exec(source);
```
[the remark this answers is not in the document]
Comment (FredKSchott) [reply packages/astro/src/vite-plugin-astro/index.ts:58]: @natemoo-re if we can't validate it inside of the compiler, what if the compiler exported a function like `validateFrontmatter` or even `validateJsExpression` so we could run it on template fragments as well? I little nervous about this being async though, if called on every template expression.

Could this be a short-term fix, and then a longer-term fix would be to add this kind of checking into the Go compiler itself later on down the road? We'll probably want the same for valid CSS...

[the remark this answers is not in the document]
Comment (drwpow) [reply packages/astro/src/vite-plugin-astro/index.ts:58]: If we follow @matthewp’s “only try on catch” strategy then this should only improve the experience. IMO the lexical scan won’t be as important because on catch we know for sure this is invalid JS. Then this would only be for error display and nothing more. So if by some happenstance that RegEx failed, it would only return a weird error message on edge cases, but in most cases would vastly improve the UX.

**Comment by drwpow** [comment]
IMO this doesn’t need a test, but if you do add one, adding a dev test in `errors.test.js` would be the easiest solution because there are very similar tests there.

File: packages/astro/src/vite-plugin-astro/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (natemoo-re) [reply packages/astro/src/vite-plugin-astro/index.ts:58]: I'm on board with that reasoning and this being the short-term fix. No objections to moving forward with Matthew's suggested approach.

File: packages/astro/test/errors.test.js [hunk 3]
```diff
@@ -21,231 +21,153 @@ before(async () => {
...
       const body = await res.text();
       console.log(res.body);
-      expect(body).to.include('unrecoverable error');
+      expect(body).to.include('Unexpected end of frontmatter');
     });
 
     it('runtime error', async () => {
-      if (isMacOS) return;
-
+      // if (isMacOS) return;
```
Comment (FredKSchott) [inline packages/astro/test/errors.test.js:43]: note: i can run these locally, so I figured I'd uncomment to see what happens. 

If they pass, I'll remove entirely.
If they fail, I'll uncomment.

Comment (FredKSchott) [reply packages/astro/test/errors.test.js:43]: update, tests are passing in CI so I'll uncomment

**Review by matthewp — APPROVED** [review]
looks great

**Review by natemoo-re — APPROVED** [review]

**Comment by FredKSchott** [comment]
works great in build too:

```
$ astro build
Error: Transform failed with 1 error:
/Users/fks/Code/astro/docs/src/pages/index.astro:4:0: error: Unexpected end of frontmatter
```
